src/Workspaces/Core/Portable/SyntaxDifferencing/SyntaxDifferenceService.cs

```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+
+namespace Microsoft.CodeAnalysis.SyntaxDifferencing
+{
+    internal abstract class SyntaxDifferenceService : ILanguageService
+    {
+        public abstract string Language { get; }
+
+        public static SyntaxDifferenceService GetService(Document document)
+        {
+            return GetService(document.Project.Solution.Workspace, document.Project.Language);
+        }
+
+        public static SyntaxDifferenceService GetService(Workspace workspace, string language)
+        {
+            return workspace.Services.GetLanguageServices(language).GetService<SyntaxDifferenceService>();
+        }
+
+        public abstract SyntaxMatch ComputeTopLevelMatch(SyntaxNode oldRoot, SyntaxNode newRoot);
```

Comment:
It's not clear if these methods need to return SyntaxMatch entirely.  Depending on what @JoshVarty needs, we could potentially expose a smaller surface area.

These functions also need explanations of what TopLevel vs BodyLevel mean.  Top-Level is simple.  it means you're passing in the root of two trees.  Body level is more subtle.  it means you're passing in the 'body' node of some piece of executable code.  i.e. something like a method body.  however, it can also mean a lambda block or the body of an expression-bodies member.  You can think about it as the entrypoint of some executable code.